tests/transaction_hooks/tests.py

```diff
@@ -0,0 +1,223 @@
+                    with atomic():
+                        self.do(1)
+                    raise ForcedError()
+            except ForcedError:
+                pass
+            self.do(2)
+
+        self.assert_done([2])
+
+    def test_no_savepoints_atomic_merged_with_outer(self):
+        with atomic():
+            with atomic():
+                self.do(1)
+                try:
+                    with atomic(savepoint=False):
+                        raise ForcedError()
+                except ForcedError:
+                    pass
+
+        self.assert_done([])
+
+    def test_inner_savepoint_does_not_affect_outer(self):
+        with atomic():
+            with atomic():
+                self.do(1)
+                try:
+                    with atomic():
+                        raise ForcedError()
+                except ForcedError:
+                    pass
+
+        self.assert_done([1])
+
+    def test_runs_hooks_in_order_registered(self):
+        with atomic():
+            self.do(1)
+            with atomic():
+                self.do(2)
+            self.do(3)
+
+        self.assert_done([1, 2, 3])
+
+    def test_hooks_cleared_after_successful_commit(self):
+        with atomic():
+            self.do(1)
+        with atomic():
+            self.do(2)
+
+        self.assert_done([1, 2])  # not [1, 1, 2]
+
+    def test_hooks_cleared_after_rollback(self):
+        try:
+            with atomic():
+                self.do(1)
+                raise ForcedError()
+        except ForcedError:
+            pass
+
+        with atomic():
+            self.do(2)
+
+        self.assert_done([2])
+
+    @skipUnlessDBFeature('test_db_allows_multiple_connections')
+    def test_hooks_cleared_on_reconnect(self):
+        with atomic():
+            self.do(1)
+            connection.close()
+
+        connection.connect()
+
+        with atomic():
+            self.do(2)
+
+        self.assert_done([2])
+
+    def test_error_in_hook_doesnt_prevent_clearing_hooks(self):
+        try:
+            with atomic():
+                connection.on_commit(lambda: self.notify('error'))
+        except ForcedError:
+            pass
+
+        with atomic():
+            self.do(1)
+
+        self.assert_done([1])
+
+    def test_db_query_in_hook(self):
+        with atomic():
+            Thing.objects.create(num=1)
+            connection.on_commit(
+                lambda: [self.notify(t.num) for t in Thing.objects.all()])
+
+        self.assert_done([1])
+
+    # On databases that don't work with autocommit off (SQLite), Atomic doesn't
+    # ever call set_autocommit(True) after committing a transaction (it sets
+    # ``connection.autocommit = True`` directly instead), so we have to run
+    # hooks immediately on commit instead of waiting for autocommit to be
+    # restored. If a hook tries to create an internal transaction of its own,
+    # this fails with an error. Need a better solution here; could be fixed if
```

Comment:
So this is an issue I'd like input on from people who know the ORM and transaction code (paging @aaugustin @akaariai and other people whose names start with A :-) ).

After committing or rolling back a transaction, `Atomic` calls `connection.set_autocommit(True)` on all database backends except SQLite (technically, all databases except those with `features.autocommits_when_autocommit_is_off`, which means SQLite). On SQLite `connection.set_autocommit(True)` is never called; instead the connection's attribute is just set directly: `connection.autocommit = True`. Because we can't hook into the latter (unless we make it a property), on SQLite we can't delay running the commit hooks until after autocommit has been turned back on, which in turn means that you can't open a transaction inside an `on_commit` hook on SQLite without Django squawking at you ("Your database backend doesn't behave properly when autocommit is off. Turn it on before using 'atomic'." - ironically in this situation autocommit is in fact on, since `Atomic` never turns it off on SQLite and the transaction has ended, but `Atomic` hasn't yet gotten around to telling the connection that it's back in autocommit mode).

We could just accept that behavior, but I think it will cause problems for people using SQLite in dev who have `on_commit` hooks that open their own transaction (this is perhaps an odd thing to do, but I'm sure some people will do it.) So I think we should fix it. Some options I see for a fix:

1) Instead of having `Atomic` poke directly at the `connection.autocommit` flag on a SQLite backend, have the backend manage it internally. That is, have the `_start_transaction_under_autocommit()` method set it to `False`, and have both `commit()` and `rollback()` set it back to `True`, only if it was originally set to `False` by entering a transaction while still in autocommit. Con: this would require a new flag to track whether autocommit is really off, or just off-because-we-started-a-transaction. Pro: this would ensure that the `autocommit` flag more accurately reflects the current state of autocommit on SQLite, eliminating the period of inaccuracy between commit/rollback and `Atomic` getting round to setting it back to `True` that causes this problem.

2) Override `set_autocommit` on the SQLite backend so that it does nothing but change the value of the flag. Pro: `Atomic` could treat all backends equally and stop worrying about `features.autocommits_when_autocommit_is_off`. Con: this is backwards-incompatible for anyone who is actually intentionally turning autocommit off on SQLite (is there any good reason why anyone would do this, given that it's broken anyway?).

3) Add a totally new API on the connection object that `Atomic` calls once a transaction is fully finished (committed or rolled back) and the connection has been fully reset to its post-transaction state (autocommit on). Something like `connection.transaction_finished()`. Then the transaction-hooks would run in this method. Pro: Doesn't require otherwise messing with the transaction/autocommit logic. Con: Adds new API of perhaps-questionable semantic value. Ties the transaction hooks even more closely to use of `Atomic` (or some other smart higher-level transaction API that knows about the need to call this method) rather than to the underlying transaction-state methods on the connection.

I think these options are listed roughly in order of my preference (though I don't have strong feelings between (1) and (2), and could be OK with any of them). It's quite likely that I'm missing some obvious reason why one or more of these wouldn't work, though. So - your thoughts?